### PR TITLE
feat: Add CLI for local Potpie usage

### LINKED ISSUE #224: Create a CLI for using potpie locally
# Create Development CLI for Local PotPie Usage

## Objective
Create a command-line interface for local development interactions with PotPie.

### 0. init 
```bash 
potpie start
``` 
and 
```bash
potpie stop
```
commands to turn on and off the server. 

### 1. parse
```bash
potpie parse <repo-path> [--branch <branch-name>]
```
- Submit repository for parsing
- Poll status until complete
- Display progress/status
- Block until ready

### 2. chat
```bash
potpie chat <repo-url> --agent <agent-name> [--branch <branch-name>]
```
- Validate project readiness
- Initialize conversation with specified agent
- Interactive console for messages
- Display responses

## Requirements
- Input validation for repo paths and branches
- Status feedback during parsing
- Error handling for invalid inputs
- Clean console output formatting
- Session management for chat

## Success Criteria
- [ ] Successfully parses repositories
- [ ] Shows parse status clearly
- [ ] Validates project status
- [ ] Handles errors gracefully

### PR DESCRIPTION
**Pull request by sofia-willow**
## Summary

Implements the CLI tool requested in #224.

### Commands

| Command | Description |
|---------|-------------|
| `potpie start` | Start Docker services + app server |
| `potpie stop` | Stop all Potpie services |
| `potpie parse <path> [--branch <name>]` | Submit repo for parsing, poll until complete |
| `potpie chat <project-id> --agent <name>` | Interactive console chat with streaming |
| `potpie status <project-id>` | Check parsing status |
| `potpie projects` | List all parsed projects |
| `potpie agents` | List available agents |

### Architecture

```
potpie/cli/
├── __init__.py
├── client.py    # HTTP client wrapping the REST API
└── main.py      # CLI commands with argparse + Rich output
```

### Design Decisions

- **Thin HTTP client** — wraps existing REST API endpoints, no direct DB access
- **Rich console output** — color-coded tables, spinners, streaming responses
- **Polling with progress** — `parse` command shows dots while waiting
- **Streaming chat** — responses stream in real-time via `httpx` streaming
- **Entry point** — added `[project.scripts] potpie = potpie.cli.main:main` to pyproject.toml
- **No new heavy dependencies** — only uses httpx (already a dep) + rich

### Tests

- `tests/cli/test_client.py` — client unit tests (health, polling, timeout)
- `tests/cli/test_main.py` — argument parsing and command dispatch tests

### Usage

```bash
# Start the server
potpie start

# Parse a repository
potpie parse /path/to/my-repo --branch main

# Check status
potpie status <project-id>

# Start chatting
potpie chat <project-id> --agent codebase_qna_agent

# List projects and agents
potpie projects
potpie agents
```

Closes #224

## Summary by CodeRabbit

* **New Features**
  * Added a user-facing CLI (potpie) to start/stop the local server, submit and monitor repo parsing, list projects and agents, view parsing status, and run interactive chat sessions with streamed assistant output; includes rich terminal formatting and a packaged CLI entry point.

* **Tests**
  * Added unit tests covering CLI argument parsing, command dispatch, and HTTP client behaviors (health, polling, auth headers).

### CODE REVIEW

**Comment by sofia-willow** [comment]
Thanks for the thorough review @coderabbitai! I've pushed fixes for all the findings:

- ✅ **Auth headers**: Now uses `x-api-key` / `x-user-id` headers (reads from `POTPIE_API_KEY` or `INTERNAL_ADMIN_SECRET` env vars)
- ✅ **agent_ids**: Fixed to use `agent_ids` list instead of `agent_id`
- ✅ **Parse timeout**: Added actual 600s timeout to prevent infinite hang
- ✅ **Status normalization**: `.lower()` before comparison
- ✅ **--branch flag**: Added to chat command
- ✅ **API prefix**: Confirmed `/api/v1` is correct per `app/main.py` router registration

Note on the API version: I verified that all routes in `app/main.py` are registered under `prefix="/api/v1"`, so the v1 prefix is correct for the current codebase.

**Comment by humblefoo02** [comment]
### PR Summary
This PR adds a CLI tool for local Potpie usage, implementing commands to start/stop services, parse repositories, check status, list projects/agents, and chat interactively with streaming responses.

| File | Changes |
|------|---------|
| potpie/cli/__init__.py | Adds package docstring |
| potpie/cli/client.py | Implements HTTP client wrapping Potpie REST API |
| potpie/cli/main.py | Implements CLI commands with argparse and Rich output |
| pyproject.toml | Adds CLI entry point and Rich dependency |
| tests/cli/__init__.py | Adds test package docstring |
| tests/cli/test_client.py | Adds HTTP client unit tests |
| tests/cli/test_main.py | Adds CLI parser/dispatch tests |

**Estimated effort:** moderate
**Potential risks:**
- Potential authentication issues if API keys aren't properly configured
- Polling logic could hang if server responses are delayed
- Streaming chat responses may have performance implications

File: potpie/cli/client.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    def __init__(
+        self,
+        base_url: str = DEFAULT_BASE_URL,
+        timeout: float = DEFAULT_TIMEOUT,
+        api_key: str | None = None,
+        user_id: str | None = None,
+    ) -> None:
+        self.base_url = base_url.rstrip("/")
+        self.timeout = timeout
+        self._api_key = api_key or os.environ.get("POTPIE_API_KEY") or os.environ.get("INTERNAL_ADMIN_SECRET") or ""
```
Comment (humblefoo02) [inline potpie/cli/client.py:36]: [warning][security] Empty string fallback for API key could mask missing auth configuration
Risk: May silently fail auth checks if neither env var is set
Suggested change: Remove empty string fallback and let ensure_auth() handle missing keys
Confidence: high

File: potpie/cli/main.py [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+            cwd=str(root),
+            check=True,
+        )
+        console.print("[green]✓ All services stopped.[/green]")
+    except subprocess.CalledProcessError as e:
+        console.print(f"[red]Error stopping services (exit {e.returncode})[/red]")
+        sys.exit(e.returncode)
+
+
+def cmd_parse(args: argparse.Namespace) -> None:
```
Comment (humblefoo02) [inline potpie/cli/main.py:130]: [warning][security] Repository path validation should handle path traversal attempts
Risk: Malicious paths could access files outside intended directory
Suggested change: Add Path(repo_path).resolve().is_relative_to(Path.cwd()) check for local paths
Confidence: high

File: potpie/cli/main.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        sys.exit(1)
+
+    console.print(f"  Project ID: [bold cyan]{project_id}[/bold cyan]")
+    console.print()
+
+    # Poll until complete (with timeout)
+    console.print("  Parsing", end="")
+    max_wait = 600.0
+    try:
+        start_time = time.monotonic()
```
Comment (humblefoo02) [inline potpie/cli/main.py:180]: [warning][correctness] Status check should handle HTTP errors explicitly
Risk: Network errors could be misinterpreted as parsing failures
Suggested change: Catch httpx.HTTPStatusError separately from other exceptions
Confidence: medium

File: potpie/cli/main.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+                continue
+
+            if user_input.strip().lower() in ("/quit", "/exit", "/q"):
+                break
+
+            # Stream response
+            console.print("[bold blue]Potpie>[/bold blue] ", end="")
+            try:
+                full_response = ""
+                for chunk in client.send_message(conversation_id, user_input):
```
Comment (humblefoo02) [inline potpie/cli/main.py:366]: [warning][reliability] Streaming response should have timeout protection
Risk: Hanging connections could block CLI indefinitely
Suggested change: Add timeout parameter to client.send_message()
Confidence: medium

File: potpie/cli/main.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+            console.print("  Wait for parsing to complete or run:")
+            console.print(f"  [bold]potpie status {project_id}[/bold]")
+            sys.exit(1)
+    except Exception as e:
+        console.print(f"[red]Error checking project status:[/red] {e}")
+        sys.exit(1)
+
+    # Create conversation
+    try:
+        conv = client.create_conversation(project_id, agent)
```
Comment (humblefoo02) [inline potpie/cli/main.py:327]: [warning][correctness] Agent validation should check against list_agents()
Risk: Invalid agent names could cause confusing errors
Suggested change: Pre-validate agent name against available agents
Confidence: high

**Comment by humblefoo02** [comment]
### PR-level observations

- **[warning/reliability]** `tests/cli/test_client.py` — Missing tests for error cases and edge conditions
  - Risk: Untested error paths could hide bugs
  - Suggested fix: Add tests for auth failures, timeouts, and malformed responses
- **[warning/reliability]** `potpie/cli/client.py` — Missing tests for error cases and edge conditions
  - Risk: Untested error paths could fail in production
  - Suggested fix: Add tests for network errors, timeouts, and malformed responses
- **[warning/reliability]** `potpie/cli/main.py` — Missing integration tests for CLI commands
  - Risk: Untested command combinations could fail in real usage
  - Suggested fix: Add end-to-end tests with mocked server responses

**Comment by humblefoo02** [comment]
### PR Review Agent summary
- Parsed findings: 28
- Posted inline findings: 8
- Severity: error=0, warning=8, info=0
- Top categories: reliability: 4, security: 2, correctness: 2
- Scope: correctness, security, reliability, cross-file impact
- Estimated effort: moderate

**Comment by humblefoo02** [comment]
### PR Summary
This PR adds a CLI tool for local Potpie usage, implementing commands to start/stop services, parse repositories, check status, list projects/agents, and chat interactively with streaming responses. The CLI wraps the existing REST API with a thin HTTP client and provides rich terminal output.

| File | Changes |
|------|---------|
| potpie/cli/__init__.py | Adds package docstring |
| potpie/cli/client.py | Implements HTTP client wrapping Potpie REST API |
| potpie/cli/main.py | Implements CLI commands with argparse and Rich output |
| pyproject.toml | Adds CLI entry point and Rich dependency |
| tests/cli/__init__.py | Adds test package docstring |
| tests/cli/test_client.py | Adds HTTP client unit tests |
| tests/cli/test_main.py | Adds CLI parser/dispatch tests |

**Estimated effort:** moderate
**Potential risks:**
- Potential authentication issues if API keys aren't properly configured
- Polling logic could hang if server responses are delayed
- Streaming chat responses may have performance implications

File: potpie/cli/client.py (hunk 1 above)
Comment (humblefoo02) [inline potpie/cli/client.py:36]: [warning][security] Long line with multiple env var fallbacks makes secret handling hard to audit
Risk: Complex secret fallback chains can hide credential issues and make debugging harder
Suggested change: Break into multiple lines with clear comments for each fallback source
Confidence: medium

File: potpie/cli/client.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        while True:
+            status = self.get_parsing_status(project_id)
+            state = status.get("status", "").lower()
+            if state in ("ready", "completed", "parsed"):
+                return status
+            if state in ("error", "failed"):
+                return status
+            if time.monotonic() - start > max_wait:
+                raise TimeoutError(
+                    f"Parsing did not complete within {max_wait}s (last status: {state})"
```
Comment (humblefoo02) [inline potpie/cli/client.py:160]: [warning][reliability] Timeout error message should include project_id for debugging
Risk: Without project_id in error, debugging failed parsing jobs is harder
Suggested change: Include project_id in timeout error message
Confidence: high

File: potpie/cli/main.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+            "or set POTPIE_ROOT."
+        )
+        sys.exit(1)
+
+    console.print(Panel.fit("🚀 Starting Potpie", style="bold green"))
+    console.print(f"  Project root: {root}")
+    console.print()
+
+    try:
+        process = subprocess.Popen(
```
Comment (humblefoo02) [inline potpie/cli/main.py:79]: [warning][security] Subprocess call uses untrusted input without validation
Risk: Potential command injection if start_script contains malicious content
Suggested change: Validate start_script path against allowlist of known scripts
Confidence: high

File: potpie/cli/main.py [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+            ["bash", str(start_script)],
```
Comment (humblefoo02) [inline potpie/cli/main.py:80]: [warning][reliability] Subprocess uses partial executable path ('bash')
Risk: Relies on PATH lookup which may resolve to unexpected binary
Suggested change: Use full path to bash executable (e.g. /bin/bash)
Confidence: medium

File: potpie/cli/main.py (hunk 2 above)
Comment (humblefoo02) [inline potpie/cli/main.py:130]: [warning][maintainability] cmd_parse function complexity exceeds threshold (13 > 10)
Risk: High complexity makes code harder to maintain and test
Suggested change: Refactor into smaller helper methods
Confidence: high

File: potpie/cli/main.py [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+
+    # Check if local path exists (skip for remote repos like owner/repo)
+    if os.path.sep in repo_path or repo_path.startswith((".", "/")):
+        local_path = Path(repo_path)
+        if not local_path.exists():
+            console.print(f"[red]Error:[/red] Local path '{repo_path}' does not exist.")
+            sys.exit(1)
+
+    if not client.is_alive():
+        console.print("[red]Error:[/red] Potpie server is not running. Run [bold]potpie start[/bold] first.")
```
Comment (humblefoo02) [inline potpie/cli/main.py:153]: [warning][correctness] Error message assumes server is always started via CLI
Risk: Misleading message if server is started via other means
Suggested change: Generalize message to not assume CLI start method
Confidence: medium

**Comment by humblefoo02** [comment]
### PR Review Agent summary
- Parsed findings: 8
- Posted inline findings: 6
- Severity: error=0, warning=6, info=0
- Top categories: security: 2, reliability: 2, maintainability: 1
- Scope: correctness, security, reliability, cross-file impact
- Merge readiness: 🚫 **HOLD** — 3 high-confidence warnings suggest non-trivial risk.
- Estimated effort: moderate

**Comment by humblefoo02** [comment]
### PR Summary
This PR adds a CLI tool for local Potpie usage, implementing commands to start/stop services, parse repositories, check status, list projects/agents, and chat interactively with streaming responses. The CLI wraps the existing REST API with a thin HTTP client and provides rich terminal output.

| File | Changes |
|------|---------|
| potpie/cli/__init__.py | Adds package docstring |
| potpie/cli/client.py | Implements HTTP client wrapping Potpie REST API |
| potpie/cli/main.py | Implements CLI commands with argparse and Rich output |
| pyproject.toml | Adds CLI entry point and Rich dependency |
| tests/cli/__init__.py | Adds test package docstring |
| tests/cli/test_client.py | Adds HTTP client unit tests |
| tests/cli/test_main.py | Adds CLI parser/dispatch tests |

**Estimated effort:** moderate
**Potential risks:**
- Potential authentication issues if API keys aren't properly configured
- Polling logic could hang if server responses are delayed
- Streaming chat responses may have performance implications

File: potpie/cli/client.py (hunk 6 above)
Comment (humblefoo02) [inline potpie/cli/client.py:160]: [warning][reliability] Timeout error message could be more actionable
Risk: Users may not know how to proceed when parsing times out
Suggested change: Include troubleshooting steps in timeout message (e.g. check server logs, try smaller repo)
Confidence: medium

File: potpie/cli/main.py [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+
+# ---------------------------------------------------------------------------
+# Helpers
+# ---------------------------------------------------------------------------
+
+def _find_project_root() -> Path:
+    """Walk up from cwd to find the Potpie project root (has compose.yaml)."""
+    current = Path.cwd()
+    for parent in [current, *current.parents]:
+        if (parent / "compose.yaml").exists() and (parent / "scripts" / "start.sh").exists():
```
Comment (humblefoo02) [inline potpie/cli/main.py:42]: [warning][correctness] Path validation could incorrectly match GitHub owner/repo format
Risk: Remote repo identifiers containing slashes (owner/repo) will be incorrectly treated as local paths
Suggested change: First check for GitHub-style owner/repo format before testing as local path
Confidence: high

File: potpie/cli/main.py (hunk 7 above)
Comment (humblefoo02) [inline potpie/cli/main.py:79]: [warning][reliability] Subprocess lacks signal handling for clean shutdown
Risk: If parent process is killed, child processes could become orphaned
Suggested change: Add signal handler to properly terminate subprocess on SIGTERM/SIGINT
Confidence: medium

File: potpie/cli/main.py [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+    stop_script = root / "scripts" / "stop.sh"
+
+    if not stop_script.exists():
+        console.print("[red]Error:[/red] Could not find scripts/stop.sh.")
+        sys.exit(1)
+
+    console.print(Panel.fit("🛑 Stopping Potpie", style="bold red"))
+
+    try:
+        subprocess.run(
```
Comment (humblefoo02) [inline potpie/cli/main.py:119]: [warning][reliability] Subprocess lacks signal handling for clean shutdown
Risk: If parent process is killed, child processes could become orphaned
Suggested change: Add signal handler to properly terminate subprocess on SIGTERM/SIGINT
Confidence: medium

File: potpie/cli/main.py (hunk 2 above)
Comment (humblefoo02) [inline potpie/cli/main.py:130]: [warning][maintainability] cmd_parse function is too complex (C901 > 10)
Risk: High complexity makes the function harder to maintain and test
Suggested change: Break into smaller helper functions for parse submission, polling, and status handling
Confidence: high

**Comment by humblefoo02** [comment]
### PR Review Agent summary
- Parsed findings: 11
- Posted inline findings: 5
- Severity: error=0, warning=5, info=0
- Top categories: reliability: 3, correctness: 1, maintainability: 1
- Scope: correctness, security, reliability, cross-file impact
- Merge readiness: ⚠️ **REVIEW** — 5 warning(s) noted. Author should address or explicitly acknowledge each.
- Estimated effort: moderate

**Comment by humblefoo02** [comment]
### PR Summary
This PR adds a CLI tool for local Potpie usage, implementing commands to start/stop services, parse repositories, check status, list projects/agents, and chat interactively with streaming responses. The CLI wraps the existing REST API with a thin HTTP client and provides rich terminal output.

| File | Changes |
|------|---------|
| potpie/cli/__init__.py | Adds package docstring |
| potpie/cli/client.py | Implements HTTP client wrapping Potpie REST API |
| potpie/cli/main.py | Implements CLI commands with argparse and Rich output |
| pyproject.toml | Adds CLI entry point and Rich dependency |
| tests/cli/__init__.py | Adds test package docstring |
| tests/cli/test_client.py | Adds HTTP client unit tests |
| tests/cli/test_main.py | Adds CLI parser/dispatch tests |

**Estimated effort:** moderate
**Potential risks:**
- Potential authentication issues if API keys aren't properly configured
- Polling logic could hang if server responses are delayed
- Streaming chat responses may have performance implications

File: potpie/cli/client.py (hunk 1 above)
Comment (humblefoo02) [inline potpie/cli/client.py:36]: [warning][security] API key fallback to empty string may cause silent auth failures
Risk: Silent auth failures make debugging harder and could lead to unexpected behavior
Suggested change: Remove empty string fallback to force explicit auth configuration
Confidence: high

Comment (humblefoo02) [inline potpie/cli/client.py:36]: [warning][maintainability] Long line exceeds PEP 8 maximum length
Risk: Reduces code readability and maintainability
Suggested change: Break line into multiple lines using parentheses
Confidence: high

File: potpie/cli/client.py (hunk 6 above)
Comment (humblefoo02) [inline potpie/cli/client.py:160]: [warning][maintainability] Long line exceeds PEP 8 maximum length
Risk: Reduces code readability and maintainability
Suggested change: Break line into multiple lines using parentheses
Confidence: high

File: potpie/cli/main.py (hunk 10 above)
Comment (humblefoo02) [inline potpie/cli/main.py:42]: [warning][correctness] Path validation could be more robust
Risk: Current validation may incorrectly match remote identifiers containing forward slashes
Suggested change: Use Path(x).exists() OR regex to detect owner/repo format first
Confidence: medium

File: potpie/cli/main.py (hunk 7 above)
Comment (humblefoo02) [inline potpie/cli/main.py:79]: [warning][reliability] Subprocess call lacks signal handler
Risk: Process may become orphaned if parent is killed
Suggested change: Add signal handler to ensure clean process termination
Confidence: medium

File: potpie/cli/main.py (hunk 8 above)
Comment (humblefoo02) [inline potpie/cli/main.py:80]: [warning][security] Subprocess call uses partial executable path
Risk: Partial paths may execute unexpected binaries if PATH is compromised
Suggested change: Use full path to bash executable
Confidence: medium

File: potpie/cli/main.py (hunk 11 above)
Comment (humblefoo02) [inline potpie/cli/main.py:119]: [warning][reliability] Subprocess call lacks signal handler
Risk: Process may become orphaned if parent is killed
Suggested change: Add signal handler to ensure clean process termination
Confidence: medium

File: potpie/cli/main.py [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+            ["bash", str(stop_script)],
```
Comment (humblefoo02) [inline potpie/cli/main.py:120]: [warning][security] Subprocess call uses partial executable path
Risk: Partial paths may execute unexpected binaries if PATH is compromised
Suggested change: Use full path to bash executable
Confidence: medium

File: potpie/cli/main.py (hunk 2 above)
Comment (humblefoo02) [inline potpie/cli/main.py:130]: [warning][maintainability] cmd_parse function is too complex
Risk: High complexity makes code harder to maintain and test
Suggested change: Refactor into smaller, focused functions
Confidence: medium

File: potpie/cli/main.py (hunk 9 above)
Comment (humblefoo02) [inline potpie/cli/main.py:153]: [warning][correctness] Error message could be more informative
Risk: Users may not understand how to fix the server not running issue
Suggested change: Include troubleshooting tips in error message
Confidence: medium

**Comment by humblefoo02** [comment]
### PR Review Agent summary
- Parsed findings: 16
- Posted inline findings: 10
- Severity: error=0, warning=10, info=0
- Top categories: security: 3, maintainability: 3, correctness: 2
- Scope: correctness, security, reliability, cross-file impact
- Merge readiness: 🚫 **HOLD** — 3 high-confidence warnings suggest non-trivial risk.
- Estimated effort: moderate

**Comment by humblefoo02** [comment]
### PR Summary
This PR adds a CLI tool for local Potpie usage, implementing commands to start/stop services, parse repositories, check status, list projects/agents, and chat interactively with streaming responses. The CLI wraps the existing REST API with a thin HTTP client and provides rich terminal output.

| File | Changes |
|------|---------|
| potpie/cli/__init__.py | Adds package docstring |
| potpie/cli/client.py | Implements HTTP client wrapping Potpie REST API |
| potpie/cli/main.py | Implements CLI commands with argparse and Rich output |
| pyproject.toml | Adds CLI entry point and Rich dependency |
| tests/cli/__init__.py | Adds test package docstring |
| tests/cli/test_client.py | Adds HTTP client unit tests |
| tests/cli/test_main.py | Adds CLI parser/dispatch tests |

**Estimated effort:** moderate
**Potential risks:**
- Potential authentication issues if API keys aren't properly configured
- Polling logic could hang if server responses are delayed
- Streaming chat responses may have performance implications

File: potpie/cli/client.py (hunk 6 above)
Comment (humblefoo02) [inline potpie/cli/client.py:160]: [warning][reliability] Timeout error message could be more actionable
Risk: Users may not know how to recover from parsing timeout
Suggested change: Add guidance on how to check parsing logs or retry
Confidence: medium

File: potpie/cli/main.py (hunk 10 above)
Comment (humblefoo02) [inline potpie/cli/main.py:42]: [warning][correctness] Path validation could incorrectly match GitHub owner/repo format
Risk: Remote repo identifiers containing forward slashes (like 'owner/repo') could be incorrectly treated as local paths
Suggested change: Add explicit check for GitHub-style owner/repo format before checking local paths
Confidence: medium

File: potpie/cli/main.py (hunk 7 above)
Comment (humblefoo02) [inline potpie/cli/main.py:79]: [warning][reliability] Subprocess could become orphaned if parent is killed
Risk: If the CLI process is killed while start.sh is running, the child process may continue running
Suggested change: Add signal handler to clean up subprocess on SIGTERM/SIGINT
Confidence: medium

File: potpie/cli/main.py (hunk 11 above)
Comment (humblefoo02) [inline potpie/cli/main.py:119]: [warning][reliability] Subprocess could become orphaned if parent is killed
Risk: If the CLI process is killed while stop.sh is running, the child process may continue running
Suggested change: Add signal handler to clean up subprocess on SIGTERM/SIGINT
Confidence: medium

File: potpie/cli/main.py (hunk 2 above)
Comment (humblefoo02) [inline potpie/cli/main.py:130]: [warning][maintainability] Function complexity exceeds recommended threshold (13 > 10)
Risk: High complexity makes the function harder to understand and maintain
Suggested change: Extract polling logic and status handling into separate helper functions
Confidence: high

File: potpie/cli/main.py (hunk 4 above)
Comment (humblefoo02) [inline potpie/cli/main.py:366]: [warning][reliability] Streaming response lacks timeout
Risk: Hanging connections could block the CLI indefinitely
Suggested change: Add timeout parameter to streaming request
Confidence: medium

**Comment by humblefoo02** [comment]
### PR Review Agent summary
- Parsed findings: 7
- Posted inline findings: 6
- Severity: error=0, warning=6, info=0
- Top categories: reliability: 4, correctness: 1, maintainability: 1
- Scope: correctness, security, reliability, cross-file impact
- Merge readiness: ⚠️ **REVIEW** — 6 warning(s) noted. Author should address or explicitly acknowledge each.
- Estimated effort: moderate

**Comment by humblefoo02** [comment]
### PR Summary
This PR adds a CLI tool for local Potpie usage, implementing commands to start/stop services, parse repositories, check status, list projects/agents, and chat interactively with streaming responses. The CLI wraps the existing REST API with a thin HTTP client and provides rich terminal output.

| File | Changes |
|------|---------|
| potpie/cli/__init__.py | Adds package docstring |
| potpie/cli/client.py | Implements HTTP client wrapping Potpie REST API |
| potpie/cli/main.py | Implements CLI commands with argparse and Rich output |
| pyproject.toml | Adds CLI entry point and Rich dependency |
| tests/cli/__init__.py | Adds test package docstring |
| tests/cli/test_client.py | Adds HTTP client unit tests |
| tests/cli/test_main.py | Adds CLI parser/dispatch tests |

**Estimated effort:** moderate
**Potential risks:**
- Potential authentication issues if API keys aren't properly configured
- Polling logic could hang if server responses are delayed
- Streaming chat responses may have performance implications

File: potpie/cli/main.py (hunk 2 above)
Comment (humblefoo02) [inline potpie/cli/main.py:130]: [warning][correctness] Path validation could incorrectly match GitHub owner/repo format
Risk: Remote repo identifiers containing forward slashes (owner/repo) could be incorrectly treated as local paths
Suggested change: First check for GitHub-style owner/repo format before testing as local path
Confidence: medium

**Comment by humblefoo02** [comment]
### PR Review Agent summary
- Parsed findings: 10
- Posted inline findings: 1
- Severity: error=0, warning=1, info=0
- Top categories: correctness: 1
- Scope: correctness, security, reliability, cross-file impact
- Merge readiness: ⚠️ **REVIEW** — 1 warning(s) noted. Author should address or explicitly acknowledge each.
- Estimated effort: moderate

**Comment by humblefoo02** [comment]
### PR Summary
This PR adds a CLI tool for local Potpie usage, implementing commands to start/stop services, parse repositories, check status, list projects/agents, and chat interactively with streaming responses. The CLI wraps the existing REST API with a thin HTTP client and provides rich terminal output.

| File | Changes |
|------|---------|
| potpie/cli/__init__.py | Adds package docstring |
| potpie/cli/client.py | Implements HTTP client wrapping Potpie REST API |
| potpie/cli/main.py | Implements CLI commands with argparse and Rich output |
| pyproject.toml | Adds CLI entry point and Rich dependency |
| tests/cli/__init__.py | Adds test package docstring |
| tests/cli/test_client.py | Adds HTTP client unit tests |
| tests/cli/test_main.py | Adds CLI parser/dispatch tests |

**Estimated effort:** moderate
**Potential risks:**
- Potential authentication issues if API keys aren't properly configured
- Polling logic could hang if server responses are delayed
- Streaming chat responses may have performance implications

File: pyproject.toml [hunk 13]
```diff
@@ -4,8 +4,13 @@ version = "0.1.0"
 description = "Add your description here"
 readme = "README.md"
 requires-python = ">=3.10"
+
```
Comment (humblefoo02) [inline pyproject.toml:7]: [error][correctness] dependencies nested under [project.scripts] — breaks pip install
Risk: pip will ignore the dependencies list entirely. Package installs without its deps.
Suggested change: Move dependencies = [...] to be a direct child of [project], not [project.scripts].
Confidence: high

File: potpie/cli/main.py [hunk 14]
```diff
@@ -0,0 +1,471 @@
...
+    if not repo_path:
+        console.print("[red]Error:[/red] Repository path cannot be empty.")
+        sys.exit(1)
+
+    if not branch:
+        console.print("[red]Error:[/red] Branch name cannot be empty.")
+        sys.exit(1)
+
+    # Check if local path exists (skip for remote repos like owner/repo)
+    if os.path.sep in repo_path or repo_path.startswith((".", "/")):
```
Comment (humblefoo02) [inline potpie/cli/main.py:146]: [error][correctness] os.path.sep in x incorrectly matches owner/repo format on Unix
Risk: Remote repo identifiers like owner/repo contain '/' which is os.path.sep on Unix, causing them to be treated as local paths and rejected as missing.
Suggested change: Check Path(x).exists() for local paths, or match owner/repo with re.match(r'^[\w.-]+/[\w.-]+$', x) first.
Confidence: high

File: potpie/cli/client.py [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+import os
+import time
+from typing import Any, Generator
+
+import httpx
+
+DEFAULT_BASE_URL = "http://localhost:8001"
+DEFAULT_TIMEOUT = 30.0
+API_PREFIX = "/api/v1"
```
Comment (humblefoo02) [inline potpie/cli/client.py:13]: [error][correctness] /api/v1 client calls but router registers /api/v2
Risk: Client and router API versions diverge (/api/v1 vs /api/v2). Calls may 404 or hit the wrong handlers at runtime.
Suggested change: Update the client hardcoded base URL/prefix to match the router prefix (/api/v2), or change the router prefix to /api/v1 intentionally.
Confidence: high
Cross-file refs: `app/main.py`

File: tests/cli/test_client.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        client = PotpieClient()
+        with pytest.raises(TimeoutError):
+            client.poll_parsing("test-id", interval=0.01, max_wait=0.05)
+
+
+class TestAPIPrefix:
+    """Test that API prefix is correctly configured."""
+
+    def test_api_prefix_value(self):
+        assert API_PREFIX == "/api/v1"
```
Comment (humblefoo02) [inline tests/cli/test_client.py:94]: [error][correctness] /api/v1 client calls but router registers /api/v2
Risk: Client and router API versions diverge (/api/v1 vs /api/v2). Calls may 404 or hit the wrong handlers at runtime.
Suggested change: Update the client hardcoded base URL/prefix to match the router prefix (/api/v2), or change the router prefix to /api/v1 intentionally.
Confidence: high
Cross-file refs: `app/main.py`

File: potpie/cli/client.py (hunk 6 above)
Comment (humblefoo02) [inline potpie/cli/client.py:160]: [warning][reliability] Timeout error message could leak internal state
Risk: Exposing internal status in error messages could aid attackers
Suggested change: Use a generic timeout message without internal status details
Confidence: medium

File: potpie/cli/main.py (hunk 7 above)
Comment (humblefoo02) [inline potpie/cli/main.py:79]: [warning][reliability] subprocess.Popen without signal handler risks orphaned processes
Risk: If parent process is killed, child process may continue running indefinitely
Suggested change: Add signal handler to ensure child process is terminated on parent exit
Confidence: high

File: potpie/cli/main.py (hunk 11 above)
Comment (humblefoo02) [inline potpie/cli/main.py:119]: [warning][reliability] subprocess.run without signal handler risks orphaned processes
Risk: If parent process is killed, child process may continue running indefinitely
Suggested change: Add signal handler to ensure child process is terminated on parent exit
Confidence: high

File: potpie/cli/main.py (hunk 2 above)
Comment (humblefoo02) [inline potpie/cli/main.py:130]: [warning][maintainability] cmd_parse function is too complex (13 > 10)
Risk: High complexity makes the function harder to maintain and test
Suggested change: Refactor into smaller helper functions
Confidence: high

File: potpie/cli/main.py [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+        start_time = time.monotonic()
+        while True:
+            status = client.get_parsing_status(project_id)
+            state = status.get("status", "").lower()
+
+            if state in ("ready", "completed", "parsed"):
+                elapsed = time.monotonic() - start_time
+                console.print(f" [green]✓ Complete![/green] ({elapsed:.0f}s)")
+                console.print()
+                console.print(f"  You can now chat with this project:")
```
Comment (humblefoo02) [inline potpie/cli/main.py:189]: [warning][maintainability] Unnecessary f-string prefix
Risk: Redundant syntax makes code harder to read
Suggested change: Remove f-string prefix since no variables are interpolated
Confidence: high

File: potpie/cli/main.py [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+                console.print(f"  [bold]potpie chat {project_id}[/bold]")
+                return
+
+            if state in ("error", "failed"):
+                console.print(f" [red]✗ Failed![/red]")
```
Comment (humblefoo02) [inline potpie/cli/main.py:194]: [warning][maintainability] Unnecessary f-string prefix
Risk: Redundant syntax makes code harder to read
Suggested change: Remove f-string prefix since no variables are interpolated
Confidence: high

**Comment by humblefoo02** [comment]
### PR-level observations

- **[warning/reliability]** `tests/cli/test_main.py` — Missing integration tests for CLI commands
  - Risk: Untested command combinations could fail in real usage
  - Suggested fix: Add end-to-end tests with mocked server responses

**Comment by humblefoo02** [comment]
### PR Review Agent summary
- Parsed findings: 13
- Posted inline findings: 11
- Severity: error=4, warning=7, info=0
- Top categories: correctness: 4, reliability: 4, maintainability: 3
- Scope: correctness, security, reliability, cross-file impact
- Merge readiness: 🚫 **HOLD** — 4 error-severity finding(s) must be resolved before merge.
- Estimated effort: moderate
